[20] 상품 정보 수정 , 삭제

ohsuha opened the pull request.

ohsuha committed:
[16] swagger 추가

ohsuha committed:
[16] swagger 추가

ohsuha committed:
[20] 상품 정보 수정 API

ohsuha committed:
[20] 상품 삭제

ohsuha committed:
[23] address 테이블 수정

ohsuha committed:
[23] db 테이블 drop 문 추가

ohsuha committed:
[23] add address

- address 관련 테이블 칼럼 수정
- 주 사용 주소는 계정당 하나씩 존재하도록

ohsuha committed:
[23] @Transactional 추가

mason1155 commented on src/main/java/org/example/commerce_site/representation/product/response/ProductResponse.java:
```diff
@@ -35,4 +35,31 @@ public static ProductResponse.Create of(ProductResponseDto.Create dto) {
+	@Getter
+	@Builder
+	@ToString
+	public static class Update {
```
Create랑 응답 포맷이 비슷한데 하나로 써도 괜찮지 않을까요??
생성, 수정할 때는 응답 안내려줘도 되긴 해요

mason1155 commented on src/main/java/org/example/commerce_site/representation/product/request/ProductRequest.java:
```diff
@@ -40,4 +40,36 @@ public static ProductRequestDto.Create toDTO(ProductRequest.Create request) {
+		@NotNull
+		//TODO : 추후 access token 에서 파싱하도록 수정
+		private Long partnerId;
+
```
한 칸씩 띄운 이유가 있나요?? 굳이 싶어서요

mason1155 commented on src/main/java/org/example/commerce_site/application/product/ProductService.java:
```diff
@@ -18,4 +20,19 @@ public class ProductService {
+		);
+	}
+
+	public Product update(Product product, ProductRequestDto.Put dto, Category category) {
```
저장, 수정, 삭제에 @Transtional 필요하지 않을까요??

ohsuha replied:
모두 수정했습니다!

ohsuha committed:
[10] yml 설정 하나로 통일

ohsuha committed:
[10] yml 설정 하나로 통일

ohsuha committed:
[20] 코드 리뷰 반영

- 불필요한 줄바꿈 삭제
- 수정시 응답 void로 변경
- 수정, 삭제에 @Transction 추가

ohsuha committed:
[20] 코드 리뷰 반영

- 불필요한 줄바꿈 삭제
- 수정시 응답 void로 변경
- 수정, 삭제에 @Transction 추가

ohsuha committed:
[20] 코드 리뷰 반영

- 불필요한 줄바꿈 삭제
- 수정시 응답 void로 변경
- 수정, 삭제에 @Transction 추가

ohsuha committed:
Merge branch 'feature/20-product-modify-delete' into feature/23-add-address

ohsuha committed:
[20] @Transactional 어노테이션 추가

ohsuha committed:
Merge pull request #24 from f-lab-edu/feature/23-add-address

[23] 배송지 추가

ohsuha committed:
Merge branch 'feature/16-swagger' into feature/20-product-modify-delete

ohsuha committed:
[16] server-url 추가

mason1155 commented:
PR 만들때 설명 적어주시면 리뷰하는데 도움됩니다 👍 
주로 적는 내용은 변경사항, 테스트 내용, 중점적으로 봐주면 좋을 곳들, 고민 포인트 등등입니다.
위 내용들을 모두 안 적으셔도 되니 일부라도 적어주세요~!

mason1155 commented on src/test/java/org/example/commerce_site/config/ModelResolverConfigTest.java:
```diff
@@ -0,0 +1,19 @@
+@Configuration
+class ModelResolverConfig {
+	@Bean
+	public ModelResolver modelResolver(ObjectMapper objectMapper) {
```
test에 ModelResolverConfig를 만든 이유가 있나요?

ohsuha replied:
이런; 잘못 만들어진것같습니다 삭제했어요!

Commit:
[20] 불필요한 코드 삭제

f-lab-kai-c commented on docs/DDL.sql:
```diff
@@ -1,3 +1,17 @@
+DROP TABLE IF EXISTS `shipments`;
```
지금 이 PR을 보기로는 스키마 정보를 DDL.sql에 모두 담아서 저장하는 것 같은데, 이러한 방법을 선택한 이유가 있는지 궁금합니다. 다른 방법으로는 어떤 형태가 있는지? 이 형태를 취했을 때의 장단점이 무엇인지?

ohsuha replied:
다른환경에서 개발을 이어갈때, api 서버를 배포할때 DDL 하나를 실행해서 서버에 필요한 디비를 생성하려고 선택했습니다. 다른 방법은 어떤것들이 있는지 일단 검색했는데 찾기가 어려운데 키워드라도 전달해 주실 수 있으실까요?ㅠ

f-lab-kai-c replied:
jpa ddl-auto나 flyway 같은 툴을 찾아보시면 좋을 것 같습니다

ohsuha replied:
flyway 적용했습니다.
1. 언제 어떤 스키마 변경이 있었는지 추적 가능
2. 비상 상황에서 이전 상태로 되돌리는 롤백 기능을 제공
3. 처음 들어보는거라서 써보고싶었습니다.

f-lab-kai-c commented on src/main/java/org/example/commerce_site/application/address/AddressService.java:
```diff
@@ -0,0 +1,27 @@
+
+	@Transactional
+	public Address createAddress(AddressRequestDto.Create dto, User user) {
+		if (dto.getIsPrimary()) {
+			addressRepository.findByUserIdAndIsPrimaryTrue(user).ifPresent(address -> {
+				address.updatePrimary();
+				addressRepository.save(address);
+			});
+		}
+		return addressRepository.save(AddressRequestDto.Create.toEntity(dto, user));
```
isPrimary인 경우에는 save가 두 번 불리게 되는 것 같은데 의도한 동작인가요?

ohsuha replied:
----엇 의도하지 않았습니다 실수입니다. 수정하겠습니다----

ohsuha replied:
죄송합니다 잘못봤습니다.. isPrimary 인 경우에는 두번 저장되는 것이 맞습니다!
주요 배송지는 하나만 등록될 수 있어서 새로 배송지를 등록할때 주요배송지로 등록하게 되면, 기존에 주요 배송지가 등록되어있다면 해제하고 새로운 배송지만 주요 배송지가 될 수 있도록 수정하는 기능입니다. 보다 명확하게 하기 위해 updatePrimary(Boolean false)를 매개변수로 받을 수 있도록 수정했습니다

f-lab-kai-c commented on src/main/java/org/example/commerce_site/application/product/ProductFacade.java:
```diff
@@ -16,8 +22,30 @@ public class ProductFacade {
+		Product product = productService.getProduct(productId);
+
+		// 파트너 본인이 올린 상품인지 확인
+		if (!Objects.equals(product.getPartnerId(), dto.getPartnerId())) {
```
여기서 `Object#equals`를 사용한 이유가 무엇인가요?

ohsuha replied:
Objects.equals 를 사용하면 혹시라도 null이 들어왔을때 방지할 수 있는 것으로 알고 있었습니다.
그런데 위의 produce, partner 를 가져오는 소스 내부에서 이미 null 일경우 예외를 던지므로 불필요해 보이네요 equals 로 수정하도록 하겠습니다.

f-lab-kai-c commented on src/main/java/org/example/commerce_site/application/product/ProductService.java:
```diff
@@ -15,7 +18,25 @@
 		return productRepository.save(ProductRequestDto.Create.toEntity(productRequest, category));
 	}
+
+	public Product getProduct(Long productId) {
```
이 메소드에는 `@Transactional` 어노테이션을 붙이지 않은 이유가 있나요?

ohsuha replied:
@Transactional은 데이터의 변경이나 생성이 있을때 사용하는 것으로 알고 있는데 get 일때는 변경이 일어나지 않으므로 붙이지 않았습니다.

f-lab-kai-c replied:
readonly 옵션에 대해서 찾아보면 좋을 것 같아요

ohsuha replied:
read only 옵션에 관해서 찾아보고 적용했습니다

f-lab-kai-c commented on src/main/java/org/example/commerce_site/config/ModelResolverConfig.java:
```diff
@@ -0,0 +1,17 @@
+public class ModelResolverConfig {
+	@Bean
+	public ModelResolver modelResolver(ObjectMapper objectMapper) {
+		return new ModelResolver(objectMapper.setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE));
```
snake_case를 활용하는 이유가 있나요?

ohsuha replied:
서버 외부로 나가는 파라미터는 모두 스네이크 케이스로 보내야한다고 들어서 스네이크 케이스로 리턴을 했습니다 그런데 RestApi가이드라는 것을 찾아보니 `JSON property에는 카멜케이스(camelCase)`를 사용하자 라는 부분이 있네요 camelCase 로 리턴 하도록 수정 하겠습니다.

f-lab-kai-c replied:
아 바꾸려면 기존 코드 전부를 바꾸셔야 할거라 바꾸진 않으셔도 될것 같습니다. 각각을 좀 찾아보시면 좋을것 같아서 코멘트 남겼습니다

f-lab-kai-c commented on src/main/java/org/example/commerce_site/domain/Address.java:
```diff
@@ -0,0 +1,41 @@
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+@SQLDelete(sql = "UPDATE addresses SET is_deleted = TRUE WHERE id = ?")
+@SQLRestriction("is_deleted IS FALSE")
```
위와 같이 soft delete를 하는 이유가 있나요?

ohsuha replied:
사용자가 이전에 주문한 내역을 불러올때 자신이 등록한 배송지에서는 삭제해 더이상 사용자에게는 보여지거나 사용할 순 없지만 이전에 주문내역을 볼때는 삭제되더라도 어떤 배송지로 발송했는지 확인할 수 있어야 하기 때문입니다

f-lab-kai-c replied:
해당 목적이라면 배송지 변경은 어떻게 대응하나요?

ohsuha replied:
배송이 시작되기 전까지 배송이 변경된다는 조건을 생각하고 있었어서 배송지 변경시에는 등록된 주소에서 새로운 주소로 변경하는 것으로 생각하고 있었습니다

f-lab-kai-c replied:
주소 A가 a를 가리키는 상태에서 배송 1을 처리하고 주소 A가 b를 가리키는 상태에서 배송 2를 처리하는 상황은 발생하지 않나요?

ohsuha replied:
헐 그 경우를 생각하지 못했네요;;;

ohsuha replied:
주소는 수정 없이 삭제와 생성만 가능하도록 해야할것 같아요 변경하겠습니다!

ohsuha committed:
[25] flyaway 적용

flyaway 선택이유
1. 명시적 마이그레이션 관리로 언제 어떤 변경이 있었는지 추적 가능
2. 롤백 가능
3. 운영환경이라고 생각하고 개발하고 싶습니다
4. 처음 보는 거라 써보고 싶어서...

ohsuha committed:
[20] updatePrimary 를 매개변수 받도록 수정

- 어떻게 업데이트를 할것인지를 매개변수 통해서 받아 메서드가 어떻게 동작하는지 명확히하고자함

ohsuha committed:
[20] transactional 어노테이션 추가

- readOnly = true 시 변경 감지를 위한 추가 작업(더티 체크)을 생략, 읽기 작업 수행시 실수로 데이터 수정하는 것 방지, 롤백을 고려할 필요 없어 커넥션이 효율적으로 관리됨의 이점에 따라 정보를 읽어오는 서비스 로직에 적용했습니다.

ohsuha committed:
Merge branch 'develop' into feature/20-product-modify-delete

f-lab-kai-c commented on src/main/java/org/example/commerce_site/application/product/ProductService.java:
```diff
@@ -15,7 +18,26 @@
+	}
+
+	@Transactional
+	public void 	delete(Product product) {
```
```suggestion
	public void delete(Product product) {
```
이런 단순한 포맷팅 실수는 일어나기 쉬운 반면, 코드 가독성은 떨어뜨립니다. 팀 단위로 작업하면 이런 실수는 더더욱 자주 일어날 텐데, 어떻게 처리해볼 수 있을까요?

ohsuha replied:
naver-intellij-formatter.xml 를 적용했는데, 커밋 전마다 잊지말고 반드시 코드 포맷팅을 실행하기...?👀

f-lab-kai-c replied:
github action 등을 활용해서 CI로 포맷팅을 검사하거나, git commit 할 때 포매팅이 돌도록 실행해보는 방법이 있을 것 같습니다

ohsuha replied:
관련해서 찾아봤는데 인텔리제이에 commit 시 자동으로 Reformat code  와 Optimize imports 해주는 옵션이 있어 이를 사용하려고 합니다
